### PR TITLE
Resolves #778: Show Severity Rating on Label

### LINKED ISSUE #778: Show Severity Rating on Label?
To help promote and reinforce the use of severity ratings, we could have them display on placed labels. Here are some mockups. Ignore actual ratings on mockups (I marked a curb ramp as a 3 but it's actually a 1 for the purposes of the mockup). Not sure I like my designs particularly as the red circle reminds me of the number of unread messages notification on my phone--and we don't want this sort of confusion.

image

image

image

Things to think about:
- Should we color the severity number (maybe map yellow to '1', orange to '3', and red to '5' to indicate growing severity--problem is that these colors could conflict with label colors).
- Should we color the severity number on the actual Likert scale on the dialog?
- Should we show the severity numbers on the minimap as well?
- Should we show more information about the label on hover (currently, hover just triggers the name of the label to show but no more information--e.g., nothing about severity or temporary status).

Semi-related to https://github.com/ProjectSidewalk/SidewalkWebpage/issues/490 (where I also mentioned showing severity ratings above labels somehow).

### PR DESCRIPTION
**Pull request by aileenzeng**
This PR features some updates to the labels.
1. Labels now display a red '?' icon when unlabeled. These go away after the user has selected a severity.
2. Label tags contain a written image + smiley corresponding to the severity rating. If there's no severity rating, it prompts the user to click the label to rate.

_Before: no display for ratings/tag just shows label type_
image

_After:_

| Severity | After |
| ------------- | ------------- |
| 1  | image | 
| 2 | image |
| 3 | image |
| 4 | image |
| 5 | image |
| unmarked | image |

### CODE REVIEW

**Comment by misaugstad** [comment]
1. After rating severity the red '?' does not go away until you move your mouse outside the label context menu (see GIF below). This isn't a dealbreaker, and I vaguely remember seeing this behavior elsewhere and thinking it looked odd. I don't know if this is an easy fix or not. I would say to look for a little while for the fix, but if you can't find anything for awhile, or the solutions you find all would be really complicated, then it isn't worth it and we can keep it as-is.
    mouse_to_remove_question_mark-3

2. Similarly, when hovering over a label, the text comes up immediately but the smiley image does not show up until you move your mouse a bit more on top of the label (see GIF below). This is a bit more problematic, since the area with the label is pretty small, and users probably aren't moving their mouse around much. But it also doesn't happen every single time! So still not a deal breaker, but since there are multiple issues like this, hopefully a solution can be found!
    mouse_to_show_smiley

This is the only issue I could find in my initial tests. Overall it looks great!! Nice work! :)

@aileenzeng Just see what you can do about the above issues and let me know. Then I'll review your code.

**Comment by misaugstad** [comment]
Well I guess I did notice that now the small videos or GIFs that show how to label during the tutorial don't show the question mark like it does in the actual tool. But I think that is very insignificant and it definitely isn't worth recording those again for such a small change.

**Comment by aileenzeng** [comment]
Thanks for testing it out! I’ll look over it today and see what I can do

**Comment by aileenzeng** [comment]
I was looking around for fixes to the first problem (not removing the '?') - it seems to be happening the labels don't re-render while the user's mouse is still inside the context menu. I tried to make the labels re-render when a user selects a severity but it didn't seem to work correctly -- it just drew a new label on top of the old one. A better fix seems to be a little gnarly.

I'm still getting some of the issues from the second problem (although less frequently now)? I changed around how some of the variables are getting stored/accessed, which has helped a little. I feel like there might be problems with `ctx.drawImage` since it seems like everything else is getting set up fine.

**Comment by misaugstad** [comment]
Hmm alright, thanks for looking into it! Can you create a new issue for it? I will review your code as-is!

**Comment by aileenzeng** [comment]
Yep - will do! Thanks :)

**Comment by jonfroehlich** [comment]
I think it’s likely crucial that the question mark disappears instantaneously as soon as severity is selected so that there is. Ice coherency here for user. Thoughts?

**Comment by aileenzeng** [comment]
I did a little more digging - I think I can actually fix the first problem (getting the question mark to disappear).

**Comment by misaugstad** [comment]
>I did a little more digging - I think I can actually fix the first problem (getting the question mark to disappear).

Nice work! Just tested it and it works great.

The smileys taking a little bit too long to show up is less important, but I agree that the question mark disappearing was pretty important. So I think this is sufficient and will review the code now!

File: public/javascripts/SVLabel/src/SVLabel/label/Label.js [hunk 1]
```diff
@@ -503,25 +509,61 @@ function Label (svl, pathIn, params) {
...
-        height = properties.tagHeight * messages.length;
-        width = -1;
+        ctx.font = '13px Open Sans';
+
+        height = properties.tagHeight * labelRows;
+
         for (i = 0; i < messages.length; i += 1) {
-            w = ctx.measureText(messages[i]).width + 5;
-            if (width < w) { width = w; }
+            // width of the tag is determined by the width of the longest row
```
Comment (misaugstad) [inline public/javascripts/SVLabel/src/SVLabel/label/Label.js]: This is suuuper nitpicky, but for comments we are trying to have comments start with a capital letter, and probably end in a period if they are more than a few words long. Mostly for the sake of consistency. I request changes on PRs for this, but I figured I should let you know at the beginning of the summer, so hopefully most of your comments end up conforming to that :)

**Comment by jonfroehlich** [comment]
How big are the smileys? They should be like 500 bytes or something... also, I recall that the pop-up images from mouseover'ing the severity Likert Scales also took a bit to load (more time than I would like--there is a noticeable lag), I wonder if this is related?

**Comment by aileenzeng** [comment]
It looks like they're ~525 bytes or so. It didn't seem to throw any
exceptions when it was drawing either?

They might be related - but I think the way the Likert scale / images in
the tag are being drawn are different too.

*Aileen Zeng*
University of Rochester '21 | Computer Science (B.S.)
azeng4@u.rochester.edu

File: public/javascripts/SVLabel/src/SVLabel/label/Label.js [hunk 2]
```diff
@@ -428,13 +430,10 @@ function Label (svl, pathIn, params) {
      * @returns {self}
      */
     function render (ctx, pov, evaluationMode) {
-        if (!evaluationMode) {
```
[the remark this answers is not in the document]
Comment (aileenzeng) [reply public/javascripts/SVLabel/src/SVLabel/label/Label.js]: I think the original code was
```
if (!evaluationMode) {
    evaluationMode = false;
}
```
I don't really remember why I made the change in the first place (this was sometime early on) so I can add it back in!

**Comment by aileenzeng** [comment]
Just made the changes! I was looking around for other places I was using `==` or `!=` - there were a few spots that I left it in when I was checking for things that were marked as `undefined` since those wouldn't correctly work with `===` or `!==`.

**Comment by misaugstad** [comment]
Looks great, congrats on your first PR!

**Comment by jonfroehlich** [comment]
Woohoo! Hi-five @aileenzeng!!!

**Pull request merged**